lib/rex/proto/dns/resolver.rb

```diff
@@ -0,0 +1,312 @@
+              rescue
+                @logger.warn "TCP Socket could not be established to #{ns}:#{@config[:port]} #{@config[:proxies]}"
+                throw :next_ns
+              end
+              next unless socket #
+              @logger.info "Contacting nameserver #{ns} port #{@config[:port]}"
+              socket.write(length+packet_data)
+              got_something = false
+              loop do
+                buffer = ""
+                ans = socket.recv(Net::DNS::INT16SZ)
+                if ans.size == 0
+                  if got_something
+                    break #Proper exit from loop
+                  else
+                    @logger.warn "Connection reset to nameserver #{ns}, trying next."
+                    throw :next_ns
+                  end
+                end
+                got_something = true
+                len = ans.unpack("n")[0]
+
+                @logger.info "Receiving #{len} bytes..."
+
+                if len.nil? or len == 0
+                  @logger.warn "Receiving 0 length packet from nameserver #{ns}, trying next."
+                  throw :next_ns
+                end
+
+                while (buffer.size < len)
+                  left = len - buffer.size
+                  temp,from = socket.recvfrom(left)
+                  buffer += temp
+                end
+
+                unless buffer.size == len
+                  @logger.warn "Malformed packet from nameserver #{ns}, trying next."
+                  throw :next_ns
+                end
+                if block_given?
+                  yield [buffer,["",@config[:port],ns.to_s,ns.to_s]]
+                else
+                  return [buffer,["",@config[:port],ns.to_s,ns.to_s]]
+                end
+              end
+            end
+  			end
+		  rescue Timeout::Error
+			  @logger.warn "Nameserver #{ns} not responding within TCP timeout, trying next one"
+			  next
+		  ensure
+			  socket.close if socket
+		  end
+		end
+		return nil
+	  end
+
+    #
+    # Send request over UDP
+    #
+    # @param packet_data [String] Data segment of DNS request packet
+    #
+    # @return ans [String] Raw DNS reply
+    def send_udp(packet_data)
+      ans = nil
+      response = ""
+      @config[:nameservers].each do |ns|
+        begin
+          @config[:udp_timeout].timeout do
+            begin
+              config = {
+                'PeerHost' => ns.to_s,
+                'PeerPort' => @config[:port].to_i,
+                'Context' => @config[:context],
+                'Comm' => @config[:comm]
+              }
+              if @config[:source_port] > 0
+                config['LocalPort'] = @config[:source_port]
+              end
+              if @config[:source_host] != IPAddr.new('0.0.0.0')
+                config['LocalHost'] = @config[:source_host] unless @config[:source_host].nil?
+              end
+              socket = Rex::Socket::Udp.create(config)
+            rescue
```

Comment:
Same concern for http://stackoverflow.com/questions/10048173/why-is-it-bad-style-to-rescue-exception-e-in-ruby